app/views/play/level/PlayLevelView.coffee

```diff
@@ -196,18 +205,53 @@ module.exports = class PlayLevelView extends RootView
         indefiniteLength: e.level.isType('game-dev')
       })
     @setupGod() if @waitingToSetUpGod
+    @levelSlug = e.level.get('slug')
+
+  checkTournamentEnd: =>
+    return unless @timeOff
+    return unless @mandate.loaded
+    return unless @levelSlug
+    courseInstanceID = @courseInstanceID or utils.getQueryVariable 'league'
+    mandate = @mandate.get('0')
+
+    tournamentState = 1
+    #        tournamentState table
+    #  ladder\checkTournamentEnd   keep    stop
+    #  open                         3       2
+    #  close                        1       0
+    if mandate
+      tournamentState = @getTournamentState mandate, courseInstanceID, @levelSlug, @timeOff
+      unless @tournamentEnd or me.isAdmin() or tournamentState > 1
+        window.location.href = '/play/ladder/'+@levelSlug+(if courseInstanceID then '/course/'+courseInstanceID else "")
+    if tournamentState % 2 == 0
+      clearInterval @checkTournamentEndInterval
+    return @checkTournamentEnd
+
+  getTournamentState: (mandate, courseInstanceID, levelSlug, timeOff) ->
+    tournament = _.find mandate.currentTournament or [], (t) =>
+      t.courseInstanceID is courseInstanceID and t.level is levelSlug
+    if tournament
+      currentTime = (Date.now() + timeOff) / 1000
+      console.log "Current time:", new Date(currentTime * 1000)
+      if currentTime < tournament.startAt
+        delta = tournament.startAt - currentTime
+        console.log "Tournament will start at: #{new Date(tournament.startAt * 1000)}, Time left: #{parseInt(delta / 60 / 60) }:#{parseInt(delta / 60) % 60}:#{parseInt(delta) % 60}"
+        return 1
+      else if currentTime > tournament.endAt
+        console.log "Tournament ended at: #{new Date(tournament.endAt * 1000)}"
+        return 0
+      delta = tournament.endAt - currentTime
+      console.log "Tournament will end at: #{new Date(tournament.endAt * 1000)}, Time left: #{parseInt(delta / 60 / 60) }:#{parseInt(delta / 60) % 60}:#{parseInt(delta) % 60}"
+      return 3
+    else
+      # 0 tournamentOnlyLevels; 2 normal ladder
+      return if levelSlug in (mandate.tournamentOnlyLevels or []) then 0 else 2
 
-    if features.china and e.level.get('slug') is 'magic-rush'
-      @checkForTournamentEnd()
-
-  checkForTournamentEnd: =>
-    return if @destroyed
-    $.get '/db/mandate', (data) =>
-      return if @destroyed
-      if data?[0]?.currentTournament isnt 'magic-rush'
-        window.location.href = '/play/ladder/magic-rush'
-      else
-        setTimeout @checkForTournamentEnd, 60 * 1000
+  calcTimeOffset: ->
+    $.ajax
+      type: 'HEAD'
+      success: (result, status, xhr) =>
+        @timeOff = new Date(xhr.getResponseHeader("Date")).getTime() - Date.now()
```

Comment:
nitpick: `timeOff` is a good shorthand but may create a slight misunderstanding of what the intent is. I think I'd prefer `timeOffset`.